[READY] Add JavaScript completion using Tern completer

puremourning opened the pull request:
## JavaScript completer

Adds support for native JavaScript completion using ycmd's Tern completer.

WIP because relies on #1848 to be merged first.

This PR:

- Adds functional documentation to the README.md
- Adds configuration documentation to the README.md
- Updates the vim doc inline with the above
- And, most importantly, updates `ycmd` submodule to include the Tern completer and the additional fixes.

## ycmd changes

There are quite a lot of commits going in to ycmd. Here's a summary of the PRs:
```
8ec8386 Auto merge of #283 - micbou:appveyor-python-issue, r=puremourning
d079a92 Auto merge of #280 - puremourning:tern-global-config, r=Valloric
dee9afe Auto merge of #281 - puremourning:tern-timeout-exit, r=Valloric
24f5d51 Auto merge of #279 - micbou:coverage, r=Valloric
115a51d Auto merge of #275 - d0k:reparse, r=Valloric
68edd9d Auto merge of #270 - micbou:refactor-tests, r=Valloric
1caabf5 Auto merge of #272 - puremourning:tern-completer-locking, r=Valloric
5c3823d Auto merge of #274 - micbou:coverage-exclude, r=vheon
22bacb3 Auto merge of #271 - puremourning:tests-coverage-doc, r=Valloric
```

Full list can be obtained via the following command in `third-party/ycmd`: `git log --oneline 22c8670c93b4ae9930c9cdf31ad9995c7fd057bd..8ec83866948b24591f62525f027b7aa4aac3e1ad`

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1849)
<!-- Reviewable:end -->

micbou commented:
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[README.md, line 1176 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BZbISOitE5Ez1N0xg:-K6BZbISOitE5Ez1N0xh:1179334869) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1176)):*
You added `javascript` in `GetParent` instead of `GetType`.

---

*[README.md, line 1207 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYVKGXWcd98HLm7sl:-K6BYVKGXWcd98HLm7sm:668041103) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1207)):*
No `GetParent` subcommand for typescript and javascript.

---

*[README.md, line 1261 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYe5M5LiJKpmXFUKI:-K6BYe5NTFJxxLc7QOqu:-2017321533) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1261)):*
Could you add `go`?

---

*[README.md, line 1268 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYoOYDiRLobk7RGMl:-K6BYoOZtLZKPwMLtALp:886361481) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1268)):*
Same comment as above.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 801 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6DOPoPD-Oo3vdrLOAP:-K6DOPoQ3xGiVjkl2RKY:798223280) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L801)):*
Or just fire a new instance of vim?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 801 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6DOPoPD-Oo3vdrLOAP:-K6DZi_ZMkPROA3NI8-J:-1040328797) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L801)):*
good shout. updated, thanks.

---

*[README.md, line 1176 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BZbISOitE5Ez1N0xg:-K6DZn5QCSyNrr2ExuIo:-109714495) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1176)):*
Doh. Thanks, fixed.

---

*[README.md, line 1207 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYVKGXWcd98HLm7sl:-K6DZpDygq29S5l02eDq:-497468049) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1207)):*
Fixed

---

*[README.md, line 1261 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYe5M5LiJKpmXFUKI:-K6DZqes-HwgOuy5-haK:-216974134) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1261)):*
bien sûr.

---

*[README.md, line 1268 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6BYoOYDiRLobk7RGMl:-K6DZxahEDkN3mweI8Io:-469593337) ([raw file](https://github.com/valloric/youcompleteme/blob/0162d4ad7157aedb9e864959e4e9910377b1cfa8/README.md#L1268)):*
done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Updated in response to @micbou and @vheon - thanks all :)

---

Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6DlmyMUD7z_KihY0hM:-761104221)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6FhuBAbZvOiFFAWlhJ:-1868698381)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Nice, thank you! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with minor comments.

After this lands, you might also want to update the YCM website by checking out the `gh-pages` branch and running `./update_from_readme.py /path/to/YouCompleteMe/README.md`. Then send a PR against the upstream `gh-pages` branch.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[README.md, line 169 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H5-FbZ4oTqICdQ6Zr:-K6H5-FbZ4oTqICdQ6Zs:-490100129) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L169)):*
Nice!

---

*[README.md, line 782 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H5EFwgtbKK5lYLPQa:-K6H5EFwgtbKK5lYLPQb:1181187452) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L782)):*
Strictly beneath or can it be the same dir? You might want to clarify that.

---

*[README.md, line 792 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H5OM2s16-L3Tc6gjF:-K6H5OM2s16-L3Tc6gjG:1703083619) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L792)):*
This sentence seems to be in some pain. :) How about this:

> YCM starts the Tern server the first time a JavaScript file is edited, so the working directory of Vim must be a descendant of the directory that contains the `.tern-project` file (or the same directory).

---

*[README.md, line 798 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H61-0GidA2M-WJ4-h:-K6H61-1EFL9hjBXmYmk:2017437064) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L798)):*
"Currently, multiple Tern servers, are not supported." -> "Multiple Tern servers are not supported."

---

*[README.md, line 800 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6APrgBceBFD36G0N:-K6H6APsBESXI40bJecN:-1070266492) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L800)):*
Either use ", or" after both the first and second bullet or skip it for both. But having it for one and not the other reads weird.

I'd go with not having ", or".

---

*[README.md, line 810 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6eQm1ndqwrIAAIe4:-K6H6eQm1ndqwrIAAIe5:-1649224943) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L810)):*
I'd skip the first sentence. It doesn't match the "voice" of the rest of the README.

---

*[README.md, line 811 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6YGVpFFcrwzt4-0d:-K6H6YGWQWkcYv57eZIJ:1716368040) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L811)):*
Make the link "Tern documentation" instead of just "documentation" because the latter makes it unclear where the link leads.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6H6xoQVyvxRUDqMOlL:1037233220)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Updated.

> you might also want to update the YCM website by checking out the gh-pages branch and running ./update_from_readme.py /path/to/YouCompleteMe/README.m

Sure thing. Will do.

I think this is ready now, just waiting on #1848 to go in first.

---

Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[README.md, line 782 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H5EFwgtbKK5lYLPQa:-K6Ikrgg214a8aptzApg:638346261) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L782)):*
Yep, same dir. Fixed, thanks.

---

*[README.md, line 792 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H5OM2s16-L3Tc6gjF:-K6IlCrI2pnJFGV0GHnp:-150764034) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L792)):*
Hah yes. Let's put it out of its misery. Much better. I think i wrinkled my nose when I wrote this, but forgot to go back and re-write it in plain English

---

*[README.md, line 798 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H61-0GidA2M-WJ4-h:-K6IleKKXtZ2Fc9JmVBx:505690559) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L798)):*
Fixed. No spoilers.

---

*[README.md, line 800 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6APrgBceBFD36G0N:-K6IlfbjISrWraedDDb-:78605006) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L800)):*
Yep, fixed. Came from changing prose to bullets :)

---

*[README.md, line 810 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6eQm1ndqwrIAAIe4:-K6Im2bLcABuOaahD99Q:985749343) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L810)):*
I changed it as follows:

> This section contains some advice for configuring `.tern-project` and working
with JavaScript files. The canonical...

The reason I put it in there is that I wanted to signpost this section as _advice_ rather than _instructions_ and didn't want to give the impression that I had any authority to speak about configuring Tern, but at the same time, based on @jwilm's comments in the ycm-dev thread, I wanted to make sure post some super-simple examples (which I think are lacking from Tern's own documentation), which I know work (because I tested them).

Perhaps a better approach would be to send a PR to Tern to add some example configuration to their docs. But I'm about 90% certain that our users would appreciate a pointer in YCM's docs, because they installed YCM, not Tern, and probably won't click that Tern documentation link.

If you're not keen on _advice_ in this section, I can move it to the wiki ?

---

*[README.md, line 811 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6YGVpFFcrwzt4-0d:-K6Imlbv7JqgjS49El5k:-84779010) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L811)):*
Sure, done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6IkMLLqrohSwpE1i7U:-47159384)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849)*
<!-- Sent from Reviewable.io -->

puremourning commented:
This should be ready now pending final signoff of the latest changes.

Merry Christmas all ;)

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6KvosAGIAunPbV7Ayb:-2108651509)*
<!-- Sent from Reviewable.io -->

vheon commented:
Let's see if @homu r+ is not too busy with Christmas eve dinner :stuck_out_tongue_closed_eyes: 

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-:-K6KyPnxmQi-eOIMxZ-b:-1130564370)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 559c762 has been approved by `vheon`

<!-- @homu r=vheon 559c76232e30b0db86cf46652e781cffdf80f56f -->

homu commented:
:hourglass: Testing commit 559c762 with merge 414f482...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/98724614)

Valloric commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[README.md, line 810 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1849#-K6H6eQm1ndqwrIAAIe4:-K6LulX9ZSaPxeIxkVMf:-688246541) ([raw file](https://github.com/valloric/youcompleteme/blob/baba751dab060e43933846bc8d539f53a5c92de7/README.md#L810)):*
I'm not at all against advice. It just that "this author" part sounded weird. Other than that one sentence which you've now replace with a superior one, all good.

Your instinct to add advice to the README is excellent.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1849)*
<!-- Sent from Reviewable.io -->

jwilm commented:
awesome work @puremourning! This is awesome :grinning:

Dsazz commented:
@puremourning Tern autocomplete works only in first call. After second call I see standart list of all js function. Why this happens ?

vheon commented:
@Dsazz minimal test case would be highly appreciated :+1:

puremourning commented:
Yeah, I'm sorry @dsazz it's hard to know exactly what the problem is without a bit more info about what's happening. I'm eager to investigate any issues asap. 

Thanks for the feedback though, much appreciated.

Dsazz commented:
@puremourning  I created new file ".tern-project" in my project:  
```javascript
{
   "libs": [
       "express",
       "...etc": "...etc"
   ],
   "plugins": {
       "node": {}
  }
}
```
And was restarted YCM server. I write for example:
```javascript
var express = require('express');
var app = express();

//First call 
app. [ list of express functions: use and etc.. ]

//Second call
app. [ list of js: constructor and etc.. ]
```

Dsazz commented:
@puremourning It's strange. This issue happens only in main app.js, but in subdirs all works normal

puremourning commented:
I will check it out. Thanks for the report.

puremourning commented:
@Dsazz how did you add `express` support to the Tern server? Did you use [tern-node-express](https://github.com/angelozerr/tern-node-express) ?

I take it you ran through the [installation instructions](https://github.com/Valloric/YouCompleteMe#installation) (i.e. adding `--tern-completer` to the YCM install) ? You don't happen to also have tern_for_vim installed ?

FWIW I'm not currently able to get Tern server to provide useful suggestions for express at all, but i'm 100% sure that's a Tern issue not a YCM one

Dsazz commented:
@puremourning  Damn. My bad. I forget about '--tern-completer'. Now all works fine and even better then early. Thanks a lot. 
P.S. I use only 'tern_for_vim'

vheon commented:
@Dsazz you don't need `tern_for_vim` now.

bpetersen commented:
What is the preferred method of installing a tern plugin such as tern-node-express with YCM --tern-completer enabled?

puremourning commented:
RE: installing tern plugins - that's a good question, and should probably be in README.md.

### Tern plugins

The Tern server supports third-party plugins. YouCompleteMe supports Tern plugins delivered as packages via `npm` and enabled via the `.tern-project` file. To install a plugin:

- Navigate to `YouCompleteMe/third_party/ycmd/third_party/tern`
- Use `npm` to install the plugin into the `node_modules` directory. See the plugin's instructions and `npm` documentation for details 
- Update your project's `.tern-project` file according to the Tern documentation and instructions supplied with the third-party plugin.

For example, to install the [`tern-node-express`](https://github.com/angelozerr/tern-node-express) plugin:

```bash
$ cd ~/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/tern
$ npm install tern-node-express
tern@0.16.1 /Users/user/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/tern
└── tern-node-express@0.3.0  extraneous
```

Then update the `.tern-project` to add the plugin:

```json
{
  "ecmaVersion": 5,
  "plugins": {
    "node": {},
    "node-express": {}
  }
}
```

That's it. You're done.

bpetersen commented:
Thanks for the info @puremourning.  I tried the method you've outlined and was not able to get tern-node-express working.  I was getting other tern information though such as parameter help for functions.

In the end, I'm using tern_for_vim and using YCM without the --tern-completer option and it seems to work great.  Plus, I'm getting the other stuff such as :TernRefs and :TernRename.

If I can help you debug, please let me know.  Also, fwiw, I'm using Neovim.

puremourning commented:
@bpetersen thanks for the info. YouCompleteMe supports the equivalent of `TernRefs` via `:YcmCompleter GotoReferences`. We don't currently support renames, but I am working on that now (I have it working using our `FixIt` infrastructure, but there are some tweaks and testing required).

On the installation of the tern plugin, strangely, I'm now not able to get it working either, though I did before when you raised the original question (I think). I'll look into it, though i'm pretty sure that it's an issue on the Tern or the 'tern-node-express' side.

puremourning commented:
I'm pretty sure this a problem with tern-node-express plugin. It expects to be installed "globally" alongside Tern in such way that it does not work properly when being installed as an npm sub-module.

From Tern documentation: 

> he main module for a third-party plugin that will be installed outside of the Tern distribution should export a single method initialize(ternDir: string) which takes the location of the Tern distribution loading the plugin as an argument. Tern will call this method immediately after loading the plugin to tell the plugin where to find Tern modules. A plugin with this method will work regardless of its install location relative to the Tern distribution, but plugins without it may fail to find Tern modules or even silently load the wrong Tern modules depending on install location.

> Plugin packages can specify the Tern version required as a peer dependency.

The tern-node-express plugin does not use the `initialise` method and so cannot load properly. It also has `tern` as a *standard* dependency. This means that you end up with another copy of `tern` in `node_modules` and it all gets messed up.

The plugin should use peer dependencies, and the `initialse` method. Peer dependencies are explained here: https://nodejs.org/en/blog/npm/peer-dependencies/.

This should be raised with the `tern-node-express` developers. Would appreciate comment from @angelozerr :) For your benefit, `ycmd` installs Tern as a git submodule, then runs `npm install --production` in that directory. We then run `npm install tern-node-express`, but this leads to another copy of `tern` in `node_modules`. Removing that leads to an error on startup:

```
module.js:328
    throw err;
    ^

Error: Cannot find module 'tern/lib/infer'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/tern/node_modules/tern-node-express/node-express.js:3:16
    at Object.<anonymous> (/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/tern/node_modules/tern-node-express/node-express.js:7:3)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
```

angelozerr commented:
@puremourning to be honnest with you, I'm not a big expert with node.

In my case I have this folder hierarchy https://github.com/angelozerr/tern.java/tree/master/core/ternjs/node_modules

If you need to update tern-node-express to fix your problem, don't hesitate to create a PR. Any contribution are welcome!

puremourning commented:
thanks @angelozerr.

Actually, on reflection I don't think my instructions above work at all. The problem is that tern plugins (even with a bunch of faffing) can't be installed as "within" the installation of tern, because they rely on tern, and `npm` isn't smart enough to know that installing `tern` within `tern` doesn't make sense. Even using peer-dependencies, they have to be installed "alongside", not "within" the tern installation. This is a pita for us. I have found a way around it but it kind of sucks.

I'm not a node expert either :/ so I will probably have to go find one. I'm sure there's one around somewhere.

Thanks again. (P.S. your tern plugins are kickass!)
